File: patterns/observer/java/Observer.java
import java.util.ArrayList;
import java.util.List;

// Observer interface
interface TemperatureObserver {
    void update(double temperature);
}

// Subject interface
interface Subject {
    void attach(TemperatureObserver observer);
    void detach(TemperatureObserver observer);
    void notifyObservers();
}

// Concrete Subject
class WeatherStation implements Subject {
    private List<TemperatureObserver> observers = new ArrayList<>();
    private double temperature = 0.0;

    @Override
    public void attach(TemperatureObserver observer) {
        observers.add(observer);
    }

    @Override
    public void detach(TemperatureObserver observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers() {
        for (TemperatureObserver observer : observers) {
            observer.update(temperature);
        }
    }

    public void setTemperature(double temp) {
        this.temperature = temp;
        notifyObservers();
    }
}

// Concrete Observers
class PhoneDisplay implements TemperatureObserver {
    private String name;

    public PhoneDisplay(String name) {
        this.name = name;
    }

    @Override
    public void update(double temperature) {
        System.out.println(name + ": Temperature updated to " + temperature + "°C");
    }
}

class WindowDisplay implements TemperatureObserver {
    private String name;

    public WindowDisplay(String name) {
        this.name = name;
    }

    @Override
    public void update(double temperature) {
        System.out.println(name + ": Temperature updated to " + temperature + "°C");
    }
}

// Main class demonstrating Observer pattern
public class Observer {
    public static void main(String[] args) {
        WeatherStation station = new WeatherStation();
        TemperatureObserver phone = new PhoneDisplay("Phone Display");
        TemperatureObserver window = new WindowDisplay("Window Display");

        station.attach(phone);
        station.attach(window);

        station.setTemperature(25.5);
        station.setTemperature(30.0);
    }
}

File: patterns/observer/python/Observer.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import List

# Observer interface
class Observer(ABC):
    @abstractmethod
    def update(self, temperature: float):
        pass

# Subject interface
class Subject(ABC):
    @abstractmethod
    def attach(self, observer: Observer):
        pass

    @abstractmethod
    def detach(self, observer: Observer):
        pass

    @abstractmethod
    def notify(self):
        pass

# Concrete Subject
class WeatherStation(Subject):
    def __init__(self):
        self._observers: List[Observer] = []
        self._temperature = 0.0

    def attach(self, observer: Observer):
        self._observers.append(observer)

    def detach(self, observer: Observer):
        self._observers.remove(observer)

    def notify(self):
        for observer in self._observers:
            observer.update(self._temperature)

    def set_temperature(self, temp: float):
        self._temperature = temp
        self.notify()

# Concrete Observers
class PhoneDisplay(Observer):
    def __init__(self, name: str):
        self.name = name

    def update(self, temperature: float):
        print(f"{self.name}: Temperature updated to {temperature:.1f}°C")

class WindowDisplay(Observer):
    def __init__(self, name: str):
        self.name = name

    def update(self, temperature: float):
        print(f"{self.name}: Temperature updated to {temperature:.1f}°C")


if __name__ == "__main__":
    station = WeatherStation()

    phone = PhoneDisplay("Phone Display")
    window = WindowDisplay("Window Display")

    station.attach(phone)
    station.attach(window)

    station.set_temperature(25.5)
    station.set_temperature(30.0)
